pkg/util/iptables/iptables.go

```diff
@@ -575,3 +576,252 @@ func IsNotFoundError(err error) bool {
+	"--name":           {true, false},
+	"--seconds":        {true, false},
+	"--dst-type":       {true, false},
+	"--set-xmark":      {true, false},
+	"--to-destination": {true, false},
+	"--mark":           {true, false},
+	"--mask":           {true, false},
+	"--mode":           {true, false},
+	"--probability":    {true, false},
+	// Argument required, can be negated
+	"-p":         {true, true},
+	"-d":         {true, true},
+	"-s":         {true, true},
+	"-i":         {true, true},
+	"-o":         {true, true},
+	"--dport":    {true, true},
+	"--src-type": {true, true},
+	// No argument required, cannot be negated
+	"--reap":    {false, false},
+	"--rsource": {false, false},
+	"--set":     {false, false},
+	// No argument required, can be negated
+	"--physdev-is-in": {false, true},
+	"--rcheck":        {false, true},
+}
+
+func parseRule(line string, allowChains []Chain, allowChainPrefixes []string) (*Rule, Chain, error) {
+	var rule Rule
+	var chain Chain
+
+	items, err := splitLine(line)
+	if err != nil {
+		return nil, "", err
+	}
+
+	for i := 0; i < len(items); i++ {
+		// negation only valid at beginning of options
+		var negated bool
+		if items[i] == "!" {
+			negated = true
+			i++
+			if i == len(items) {
+				return nil, "", fmt.Errorf("invalid iptables rule (not enough arguments): %v", line)
+			}
+		}
+
+		opt := items[i]
+		attr, ok := optionAttrs[opt]
+		if !ok {
+			return nil, "", fmt.Errorf("Unknown iptables rule option %q", opt)
+		}
+		if !attr.canNegate && negated {
+			return nil, "", fmt.Errorf("invalid iptables rule (can't negate option %q): %v", opt, line)
+		}
+
+		if attr.argRequired {
+			i++
+			if i == len(items) {
+				return nil, "", fmt.Errorf("invalid iptables rule (not enough arguments): %v", line)
+			}
+
+			// Validate/process the required argument
+			switch opt {
+			case "-A":
+				if chain != "" {
+					return nil, "", fmt.Errorf("invalid iptables rule (multiple \"-A\" args): %v", line)
+				}
+				chain = Chain(items[i])
+				if !isChainAllowed(chain, allowChains, allowChainPrefixes) {
+					return nil, "", nil
+				}
+			case "-j":
+				if rule.Jump != "" {
+					return nil, "", fmt.Errorf("invalid iptables rule (multiple \"-j\" args): %v", line)
+				}
+				rule.Jump = Chain(items[i])
+			case "--comment":
+				rule.Comment = items[i]
+			case "-m":
+				rule.Modules = append(rule.Modules, items[i])
+			default:
+				if err := addRuleOption(&rule, opt, RuleOption{Arg: items[i], Negated: negated}); err != nil {
+					return nil, "", err
+				}
+			}
+		} else {
+			if err := addRuleOption(&rule, opt, RuleOption{Negated: negated}); err != nil {
+				return nil, "", err
+			}
+		}
+	}
+
+	// Some post fixups for default module options that will show up
+	// in iptables-save output, which which may not be added when constructing
+	// rules manually
+	for _, module := range rule.Modules {
+		switch module {
+		case "recent":
+			if _, ok := rule.Options["--rsource"]; !ok {
```

Comment:
I don't understand these?  It does raise another issue - there are many things that the iptables command allows that get mutated on readback.  --mask becomes --xmask, and the hex value of the mask gets canonicalized.  If we can't handle that all, I am reluctant to embed this logic in a generic location - people will assume it works.

kube-proxy pre-canonicalizes mostly for historical reasons, but it would be really easy to slip up on that (or to have drift across tool versions).  Since this is distant from the use-site, it will be very hard to find.